### PR TITLE
[dev] Enable cudnn fused kernel in dsa for DeepSeek-V3.2

### PR DESCRIPTION
**Pull request by kunlunl**
# What does this PR do ?

Enabled cudnn fused DSA kernels that released in https://github.com/NVIDIA/cudnn-frontend release v1.24.0

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 1]
```diff
@@ -1181,85 +739,97 @@ def forward(
...
+
+        sq, b, np, _ = query.size()
         skv = key.size(0)
-        hnv = value.size(3)
+        d_v = self.v_channels or self.config.kv_lora_rank or self.config.v_head_dim
+        assert key.size(-1) >= d_v, "DSAttention key must contain the value channels."
+
+        backend = getattr(self.config, "attention_backend", None)
+        indexer_sparse_loss = getattr(self.config, "dsa_indexer_use_sparse_loss", True)
+        use_unfused = (
```
Comment (hxbai) [inline megatron/core/transformer/experimental_attention_variant/dsa.py]: I added `apply_dsa_kernel_fusion` in transformer_config. Maybe you can use that flag.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 2]
```diff
@@ -1181,85 +739,97 @@ def forward(
...
+        assert key.size(-2) == 1, "DSAttention expects MQA KV with one KV head."
+
+        sq, b, np, _ = query.size()
         skv = key.size(0)
-        hnv = value.size(3)
+        d_v = self.v_channels or self.config.kv_lora_rank or self.config.v_head_dim
+        assert key.size(-1) >= d_v, "DSAttention key must contain the value channels."
+
+        backend = getattr(self.config, "attention_backend", None)
+        indexer_sparse_loss = getattr(self.config, "dsa_indexer_use_sparse_loss", True)
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Fixed. The fallback now matches the TransformerConfig default (`False`) instead of changing the behavior for configs that do not carry the field.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 3]
```diff
@@ -717,24 +723,20 @@ def _forward_unfused_csa(
...
                         k_indexer,
+                        weights_indexer,
+                        self.indexer.softmax_scale,
                         query.detach(),
                         key_for_loss.detach(),
                         self.softmax_scale,
                         min(self.indexer.index_topk, n_compressed),
                         indexer_loss_coeff,
-                        causal_mask,
                         getattr(self.config, "dsa_indexer_use_sparse_loss", True),
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: Fixed as well. Now use `False` as the fallback to stay consistent with TransformerConfig.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 4]
```diff
@@ -1181,85 +739,97 @@ def forward(
...
+                mask=float_mask,
             )
-            # Save indexer loss for logging
-            if indexer_loss_coeff > 0:
-                DSAIndexerLossLoggingHelper.save_loss_to_tracker(
-                    loss=indexer_loss,
-                    layer_number=self.layer_number,
-                    num_layers=self.config.num_layers + (self.config.mtp_num_layers or 0),
-                )
+            output = unfused_dsa_fn(query, key, topk_indices, self.softmax_scale, d_v)
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:846]: Fixed. The unfused path now computes the KL/indexer loss only when training with grad enabled and a positive loss coeff; eval/no_grad only computes the top-k indices.

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 5]
```diff
@@ -474,30 +480,70 @@ def _compute_indexer_predict(
...
+        from cudnn.deepseek_sparse_attention.score_recompute import _interface_sm90
+
+        return _interface_sm90.sparse_indexer_score_recompute(
+            q_indexer_bshd,
+            k_indexer_bsd,
+            weights_bsh,
+            topk_indices,
+            sm_scale=indexer_softmax_scale,
+            topk_indices_global=False,
+        )
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py:530]: Added a comment. The SM90 cuDNN entry point does not expose `qhead_per_kv_head`, while the SM100 entry point does, so the signatures intentionally differ.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -841,6 +491,10 @@ def __init__(
         if pg_collection is None:
             pg_collection = ProcessGroupCollection.use_mpu_process_groups(required_pgs=['tp', 'cp'])
         self.pg_collection = pg_collection
+        if getattr(self.config, "tensor_model_parallel_size", 1) != 1:
+            assert False, "Tensor parallelism is not supported for DSAIndexer."
+        if self.pg_collection.tp.size() > 1:
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:496]: Fixed. Replaced the unconventional `assert False` pattern with `raise ValueError(...)` for the unsupported TP checks.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 1 above)
Comment (kunlunl) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Thanks. Now use `apply_dsa_kernel_fusion` to decide whether to take the fused DSA path, together with the attention backend. The fused parity test enables this flag explicitly as well.

**Review by hxbai — APPROVED** [review]
LGTM

**Comment by hxbai** [comment]
There is some related work in the main branch. Close this for now.
